File: BasicMetadataHandling/src/main/java/com/bitmovin/player/samples/metadata/basic/MainActivity.java
/*
 * Bitmovin Player Android SDK
 * Copyright (C) 2017, Bitmovin GmbH, All Rights Reserved
 *
 * This source code and its use and distribution, is subject to the terms
 * and conditions of the applicable license agreement.
 */

package com.bitmovin.player.samples.metadata.basic;

import android.os.Bundle;
import android.util.Log;

import androidx.appcompat.app.AppCompatActivity;

import com.bitmovin.player.PlayerView;
import com.bitmovin.player.api.Player;
import com.bitmovin.player.api.event.EventListener;
import com.bitmovin.player.api.event.PlayerEvent;
import com.bitmovin.player.api.event.SourceEvent;
import com.bitmovin.player.api.metadata.Metadata;
import com.bitmovin.player.api.metadata.emsg.EventMessage;
import com.bitmovin.player.api.metadata.id3.Id3Frame;
import com.bitmovin.player.api.metadata.scte.ScteMessage;
import com.bitmovin.player.api.source.SourceConfig;
import com.google.gson.Gson;

public class MainActivity extends AppCompatActivity {

    private PlayerView playerView;
    private Player player;
    private Gson gson;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        this.playerView = this.findViewById(R.id.bitmovinPlayerView);
        this.player = this.playerView.getPlayer();

        this.gson = new Gson();
        // Adding the metadata listener to the player
        this.player.on(SourceEvent.MetadataParsed.class, metadataParsedListener);
        this.player.on(PlayerEvent.Metadata.class, metadataListener);

        this.initializePlayer();
    }

    @Override
    protected void onStart() {
        this.playerView.onStart();
        super.onStart();
    }

    @Override
    protected void onResume() {
        super.onResume();
        this.playerView.onResume();
    }

    @Override
    protected void onPause() {
        this.playerView.onPause();
        super.onPause();
    }

    @Override
    protected void onStop() {
        this.playerView.onStop();
        super.onStop();
    }

    @Override
    protected void onDestroy() {
        // Removing metadata listener from player
        this.player.off(metadataParsedListener);
        this.player.off(metadataListener);
        this.playerView.onDestroy();

        super.onDestroy();
    }

    protected void initializePlayer() {
        //TODO: add source containing metadata
        this.player.load(SourceConfig.fromUrl(""));
    }

    private void logMetadata(Metadata metadata, String type) {
        switch (type) {
            case ScteMessage.TYPE:
                for (int i = 0; i < metadata.length(); i++) {
                    ScteMessage scteMessage = (ScteMessage) metadata.get(i);

                    Log.i("METADATA", "SCTE: " + gson.toJson(scteMessage));
                }
                break;
            case Id3Frame.TYPE:
                for (int i = 0; i < metadata.length(); i++) {
                    Id3Frame id3Frame = (Id3Frame) metadata.get(i);

                    Log.i("METADATA", "ID3Frame: " + gson.toJson(id3Frame));
                }
                break;
            case EventMessage.TYPE:
                for (int i = 0; i < metadata.length(); i++) {
                    EventMessage eventMessage = (EventMessage) metadata.get(i);

                    Log.i("METADATA", "EMSG: " + gson.toJson(eventMessage));
                }
                break;
        }
    }

    // Metadata Listener
    private final EventListener<PlayerEvent.Metadata> metadataListener = metadata -> {
        Log.i("METADATA", "onMetadata:");
        logMetadata(metadata.getMetadata(), metadata.getType());
    };

    // Parsed Metadata Listener
    private final EventListener<SourceEvent.MetadataParsed> metadataParsedListener = metadataParsed -> {
        Log.i("METADATA", "onMetadataParsed:");
        logMetadata(metadataParsed.getMetadata(), metadataParsed.getType());
    };
}

File: BasicMetaDataHandlingKotlin/src/main/java/com/bitmovin/player/samples/metadata/basic/MainActivity.kt
package com.bitmovin.player.samples.metadata.basic

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import android.util.Log
import com.bitmovin.player.api.Player
import com.bitmovin.player.api.event.PlayerEvent
import com.bitmovin.player.api.event.SourceEvent
import com.bitmovin.player.api.event.on
import com.bitmovin.player.api.metadata.emsg.EventMessage
import com.bitmovin.player.api.metadata.id3.Id3Frame
import com.bitmovin.player.api.metadata.scte.ScteMessage
import com.bitmovin.player.api.source.SourceConfig
import com.google.gson.Gson
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    private lateinit var player: Player
    private val gson: Gson = Gson()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        player = bitmovinPlayerView.player!!

        // Adding the metadata listener to the player
        player.on(::onMetadataParsed)
        player.on(::onMetadata)

        initializePlayer()
    }

    override fun onStart() {
        bitmovinPlayerView.onStart()
        super.onStart()
    }

    override fun onResume() {
        super.onResume()
        bitmovinPlayerView.onResume()
    }

    override fun onPause() {
        bitmovinPlayerView.onPause()
        super.onPause()
    }

    override fun onStop() {
        bitmovinPlayerView.onStop()
        super.onStop()
    }

    override fun onDestroy() {
        // Removing metadata listener from player
        player.off(::onMetadataParsed)
        player.off(::onMetadata)
        bitmovinPlayerView.onDestroy()

        super.onDestroy()
    }

    private fun initializePlayer() {
        //TODO: add source containing metadata

        // load source using the created source item
        player.load(SourceConfig.fromUrl(""))
    }

    private fun logMetadata(metadata: com.bitmovin.player.api.metadata.Metadata, type: String) {
        when (type) {
            ScteMessage.TYPE -> (0 until metadata.length())
                    .map { metadata.get(it) as ScteMessage }
                    .forEach { Log.i("METADATA", "SCTE: " + gson.toJson(it)) }
            Id3Frame.TYPE -> (0 until metadata.length())
                    .map { metadata.get(it) as Id3Frame }
                    .forEach { Log.i("METADATA", "ID3Frame: " + gson.toJson(it)) }
            EventMessage.TYPE -> (0 until metadata.length())
                    .map { metadata.get(it) as EventMessage }
                    .forEach { Log.i("METADATA", "EMSG: " + gson.toJson(it)) }
        }
    }

    private fun onMetadataParsed(event: SourceEvent.MetadataParsed){
        Log.i("METADATA", "onMetadataParsed:")
        logMetadata(event.metadata, event.type)
    }

    private fun onMetadata(event: PlayerEvent.Metadata){
        Log.i("METADATA", "onMetadata:")
        logMetadata(event.metadata, event.type)
    }

}
